### PR TITLE
ICU-20568 Improve MacroProps error handling

### PR DESCRIPTION
**Pull request by hugovdm**
In particular: actually handle Usage memory allocation errors (Closes: https://github.com/icu-units/icu/issues/89).
Also: correct Scale's error condition.

##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20568
- [X] Updated PR title and link in previous line to include Issue number
- [X] Issue accepted
- [ ] Tests included
- [ ] Documentation is changed or added

### CODE REVIEW

**Comment by hugovdm** [comment]
As of b843d9c it looks like I've only got an error for [osx clang](https://travis-ci.org/github/unicode-org/icu/jobs/726332123):

```
...
pkgdata: ar r -c uconvmsg/libuconvmsg.a uconvmsg/uconvmsg_dat.o
pkgdata: ar r -c uconvmsg/libuconvmsg.a uconvmsg/uconvmsg_dat.o
pkgdata: ranlib uconvmsg/libuconvmsg.a
/Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: archive member: uconvmsg/libuconvmsg.a() size too large (archive member extends past the end of the file)
-- return status = 256
Error generating library file. Failed command: ranlib uconvmsg/libuconvmsg.a
Error generating assembly code for data.
make[2]: *** [uconvmsg/libuconvmsg.a] Error 1
make[1]: *** [all-recursive] Error 2
make[1]: *** Waiting for unfinished jobs....
...
```

**Comment by hugovdm** [comment]
I squashed yesterday, partially to encourage the CI tests to rerun. All checks pass. Please review?

File: icu4c/source/i18n/unicode/numberformatter.h [hunk 1]
```diff
@@ -1174,10 +1173,24 @@ class U_I18N_API Usage : public UMemory {
...
     friend class impl::GeneratorHelpers;
 
-    // Allow MacroProps/MicroProps to initialize empty instances.
-    friend struct impl::MacroProps;
-
 #endif // U_HIDE_INTERNAL_API
+
+  private:
+    /** @internal */
+    UErrorCode fError;
```
Comment (sffc) [inline icu4c/source/i18n/unicode/numberformatter.h:1180]: Please keep fields with fields, functions with functions, and friends with friends.  Also, remove the second redundant `private:`. If `#endif // U_HIDE_INTERNAL_API` is causing problems, just move it up above the first `private:`.

**Review by sffc — CHANGES_REQUESTED** [review]

File: icu4c/source/i18n/unicode/numberformatter.h (hunk 1 above)
Comment (hugovdm) [reply icu4c/source/i18n/unicode/numberformatter.h]: Done. Not squashed yet.

**Review by sffc — COMMENTED** [review]
LGTM please squash

**Review by sffc — APPROVED** [review]

**Pull request merged**